### ISSUE TITLE
Kernel crash at tempesta_fw/server.c:39 when Tempesta is stopped

### ISSUE THREAD

**Issue by keshonok** [natsys/tempesta#127]
Different circumstances. It could be an attempts to stop a running Tempesta with a stop command. Or it could be Tempesta stopping when it encountered an unrecoverable error while loading (on start command). The common part of the trace is below.

```
[    8.603159] [tempesta]   stopping already stared modules
[    8.603637] [tempesta]   mod_stop(): sock_srv
[    8.604061] [tempesta]   Free msg:           (null)
[    8.604510] [tempesta]   Free msg:           (null)
[    8.604936] [tempesta]   Free msg:           (null)
[    8.605373] [tempesta]   Free msg:           (null)
[    8.605808] [tempesta]   mod_stop(): http_sticky
[    8.606221] [tempesta]   mod_stop(): cache
[    8.824699] [tdb] Close table cache
[    8.825051] [tempesta]   mod_stop(): filter
[    8.841880] [tdb] Close table filter
[    8.842253] [tempesta]   spec cleanup: 'listen'
[    8.842673] [tempesta]   spec cleanup: 'server'
[    8.843115] ------------[ cut here ]------------
[    8.843537] kernel BUG at /home/natsys-lab/tfw-run/tempesta/tempesta_fw/server.c:39!
[    8.844018] invalid opcode: 0000 [#1] SMP 
[    8.844018] Modules linked in: tfw_sched_rr(O) tfw_sched_http(O) tfw_sched_hash(O) tfw_sched_dummy(O) tempesta_fw(O) tempesta_db(O)
[    8.844018] CPU: 0 PID: 2074 Comm: sysctl Tainted: G        W  O 3.10.10-tempesta #2
[    8.844018] task: ffff8800389b6100 ti: ffff880038c22000 task.ti: ffff880038c22000
[    8.844018] RIP: 0010:[<ffffffffa004cd22>]  [<ffffffffa004cd22>] tfw_destroy_server+0x25/0x3f [tempesta_fw]
[    8.844018] RSP: 0018:ffff880038c23c90  EFLAGS: 00010202
[    8.844018] RAX: 0000000000000001 RBX: ffff880038c23df0 RCX: a000000000000000
[    8.844018] RDX: ffffffffa005bff0 RSI: 1268000000000000 RDI: ffffffffa005bf88
[    8.844018] RBP: ffff880038c23c98 R08: ffffffff81acb070 R09: 0000000000000000
[    8.844018] R10: 0000000000000000 R11: 0000000000000003 R12: 000000000000001f
[    8.844018] R13: 0000000000000000 R14: 000000000000001f R15: 0000000000000000
[    8.844018] FS:  00007f870f857700(0000) GS:ffff88003fc00000(0000) knlGS:0000000000000000
[    8.844018] CS:  0010 DS: 0000 ES: 0000 CR0: 000000008005003b
[    8.844018] CR2: 00007f870f0e48b0 CR3: 0000000038d2a000 CR4: 00000000000006b0
[    8.844018] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[    8.844018] DR3: 0000000000000000 DR6: 00000000ffff0ff0 DR7: 0000000000000400
[    8.844018] Stack:
[    8.844018]  ffffffffa005bf88 ffff880038c23cf8 ffffffffa004d382 ffffffffa005bf88
[    8.844018]  ffff880038aebed8 ffff880038aebf40 ffffffffa005bff0 ffff880038aebf40
[    8.844018]  ffffffffa005bff0 ffffffffa005bff0 ffff880038aebf40 ffffffffa00516f8
[    8.844018] Call Trace:
[    8.844018]  [<ffffffffa004d382>] tfw_sg_release_all+0x8c/0x151 [tempesta_fw]
[    8.844018]  [<ffffffffa00516f8>] ? tfw_sock_srv_delete_all_conns+0x14/0x2a [tempesta_fw]
[    8.844018]  [<ffffffffa0051bc6>] tfw_srv_cfg_clean_srv_groups+0x16/0x23 [tempesta_fw]
[    8.844018]  [<ffffffffa000f8f8>] spec_cleanup+0x82/0xa4 [tempesta_fw]
[    8.844018]  [<ffffffffa0010b07>] tfw_cfg_start_mods+0x1c7/0x1ea [tempesta_fw]
[    8.844018]  [<ffffffffa0010f2d>] handle_state_change+0xfd/0x192 [tempesta_fw]
[    8.844018]  [<ffffffffa0011111>] handle_sysctl_state_io+0x14f/0x1ba [tempesta_fw]
[    8.844018]  [<ffffffffa0011016>] ? handle_sysctl_state_io+0x54/0x1ba [tempesta_fw]
[    8.844018]  [<ffffffff811408bf>] proc_sys_call_handler+0x8f/0xc0
[    8.844018]  [<ffffffff811408bf>] ? proc_sys_call_handler+0x8f/0xc0
[    8.844018]  [<ffffffff811408ff>] proc_sys_write+0xf/0x20
[    8.844018]  [<ffffffff810e91b3>] vfs_write+0xd3/0x1b0
[    8.844018]  [<ffffffff810e9787>] SyS_write+0x47/0x90
[    8.844018]  [<ffffffff813cb2a9>] system_call_fastpath+0x16/0x1b
[    8.844018] Code: b5 21 1a e1 c9 c3 55 48 89 e5 48 83 ec 08 48 89 7d f8 48 8b 45 f8 48 89 c7 e8 71 ff ff ff 85 c0 0f 94 c0 0f b6 c0 48 85 c0 74 02 <0f> 0b 48 8b 05 fd 77 01 00 48 8b 55 f8 48 89 d6 48 89 c7 e8 46 
[    8.844018] RIP  [<ffffffffa004cd22>] tfw_destroy_server+0x25/0x3f [tempesta_fw]
[    8.844018]  RSP <ffff880038c23c90>
[    8.868808] ---[ end trace d25308c7fbbccc35 ]---
```

**Comment by keshonok** [comment]
This is a more specific bug report. The original issue is #92.

**Comment by keshonok** [comment]
Fixed in cd60717
